### PR TITLE
test(IC-1579): Isolate TLA tests in Rust-based (non-canister) tests

### PR DESCRIPTION
**Pull request by oggy-dfin**
Switch to a `LocalKey`-based approach for Rust-based tests, and add an annotation (proc macro) that scopes the `LocalKey` to the test. 

Previously, we used only a global `RwLock`, which would cause traces of one test to be picked up by other tests. In contract, a `LocalKey` is task-local, eliminating the interference. We still use the `RwLock` as a fallback to store traces if the `LocalKey` is not present (i.e., no scope has been opened). This means that, first, tests that aren't annotated still collect the traces, but they aren't checked, which is a bit wasteful but should be OK as these aren't extremely heavy operations. Second, and more importantly, this means that the instrumentation will still work in canister-based tests, allowing a separate query method to pick the traces up from the global variable.

### CODE REVIEW

File: rs/nns/governance/src/governance/tla/mod.rs [hunk 1]
```diff
@@ -213,10 +213,25 @@ pub fn check_traces() {
     // improving that later, for now we introduce a hard limit on the state size, and
     // skip checking states larger than the limit
     const STATE_SIZE_LIMIT: u64 = 500;
+    fn is_under_limit(p: &ResolvedStatePair) -> bool {
```
Comment (oggy-dfin) [inline rs/nns/governance/src/governance/tla/mod.rs:216]: This block is not really related to the change, it's just there to print some stats related to the state size filter introduced in #2188

File: rs/nns/governance/src/governance/tla/mod.rs [hunk 2]
```diff
@@ -267,7 +284,7 @@ pub fn check_traces() {
                 println!("Possible divergence from the TLA model detected when interacting with the ledger!");
                 println!("If you did not expect to change the interaction between governance and the ledger, reconsider whether your change is safe. You can find additional data on the step that triggered the error below.");
                 println!("If you are confident that your change is correct, please contact the #formal-models Slack channel and describe the problem.");
-                println!("You can edit nervous_system/tla/feature_flags.bzl to disable TLA checks in the CI and get on with your business.");
+                println!("You can edit nns/governance/feature_flags.bzl to disable TLA checks in the CI and get on with your business.");
```
Comment (oggy-dfin) [inline rs/nns/governance/src/governance/tla/mod.rs:287]: Unrelated fix for a stale instruction

File: rs/nns/governance/src/governance.rs [hunk 3]
```diff
@@ -135,7 +135,7 @@ pub mod tla;
 #[cfg(feature = "tla")]
 pub use tla::{
     claim_neuron_desc, split_neuron_desc, tla_update_method, InstrumentationState, ToTla,
-    TLA_INSTRUMENTATION_STATE, TLA_TRACES,
```
Comment (max-dfinity) [inline rs/nns/governance/src/governance.rs:138]: I don't see where these are used, but I assume it is from macro expansion?

**Review by max-dfinity — APPROVED** [review]

File: rs/nns/governance/src/governance.rs (hunk 3 above)
Comment (oggy-dfin) [reply rs/nns/governance/src/governance.rs:138]: Yep, the macros assume these variables to be in scope (I added some doc comments to the macros)

**Review by mraszyk — APPROVED** [review]
